fix(amf): Unit test cases for PDUSession Establishment/Release

ganeshg87 opened the pull request:
Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

Unit test cases for PDUSession Establishment/Release

## Test Plan

 verified memory leaks & code coverage for the below Message Handling
-  PDU Session Establishment Request Handling by AMF
-   PDU Session Release Handling by AMF
        ngap - amf
	amf  - Pipelined
	amf  - smf
- Registration and PDUSession Establishment/Release with UERANSIM
![image](https://user-images.githubusercontent.com/83060027/141168365-65f57c12-7095-46ff-85cf-35d2d942ee1d.png)
[mme_reg_pdu_rel_10_nov_21.log](https://github.com/magma/magma/files/7514965/mme_reg_pdu_rel_10_nov_21.log)

- executed test_oai
![image](https://user-images.githubusercontent.com/83060027/141168170-e468d29e-a2b2-4807-837c-ef6d8ad4a68f.png)

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

ganeshg87 committed:
fix(amf): Unit test cases for PDUSession Establishment/Release

Test cases:
  -verified memory leaks & code coverage for below Message Handling
    PDU Session Establishment Request Handling by AMF
    PDU Session Release Handling by AMF
        ngap - amf
	amf  - Pipelined
	amf  - smf

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:
```diff
@@ -617,13 +616,13 @@ void amf_app_handle_pdu_session_response(
           LOG_AMF_APP, "pdu session  not found for session_id = %u\n",
           pdu_session_resp->pdu_session_id);
-      return;
+      return rc;
```
The assumption here is that `rc` is initialized to RETURNerror and has not been modified since. It is better to be more explicit.
```suggestion
      return RETURNerror;
```

ganeshg87 replied:
addressed the review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:
```diff
@@ -617,13 +616,13 @@ void amf_app_handle_pdu_session_response(
     OAILOG_ERROR(
         LOG_AMF_APP, "ue context not found for the imsi=%lu\n", imsi64);
-    return;
+    return rc;
```
Better to be explicit
```suggestion
    return RETURNerror;
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp:
```diff
@@ -15,6 +15,9 @@
 #include "M5GAuthenticationServiceClient.h"
 #include "amf_common.h"
 #include <memory>
+#include "lte/gateway/c/core/oai/lib/mobility_client/MobilityServiceClient.h"
```
Is this needed here?

ganeshg87 replied:
addressed review comment, removed header file

ssanadhya replied:
@ganeshg87 , did you forget to push the commit? The header is still included.

ulaskozat replied:
@ganeshg87 looks like you did not address the comment?

ganeshg87 replied:
@ssanadhya, @ulaskozat,
   It was my mistake, the review comment was addressed now.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h:
```diff
@@ -60,10 +84,29 @@ class AMFClientServicer : public AMFClientServicerBase {
+        bdestroy(session_transfer->up_transport_layer_info.gtp_tnl
+                     .endpoint_ip_address);
+        bdestroy(pdusession_resource_setup_req->nas_pdu);
+        rc = RETURNok;
```
Add this clean up to the definition of `itti_free_msg_content` . Further, bdestroy_wrapper is preferred over bdestroy.

ganeshg87 replied:
addressed the comment, moved cleanup code under itti_free_msg_content

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/prepare_request_for_smf.cpp:
```diff
@@ -92,7 +93,11 @@ int create_session_grpc_req_on_gnb_setup_rsp(
 
-  AsyncSmfServiceClient::getInstance().set_smf_session(req);
+  OAILOG_INFO(
+      LOG_AMF_APP, "Sending msg(grpc) to :[sessiond] for ue: [%s] session\n",
```
```suggestion
      LOG_AMF_APP, "Sending msg(grpc) to [sessiond] for ue: [%s] \n",
```

ganeshg87 replied:
addressed the comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,212 @@ int send_uplink_nas_registration_complete(
+  bstring pdu_session_est_req;
+  tai_t originating_tai = {};
+
+  if ((!amf_app_desc_p) || (!nas_msg) || (0 == nas_msg_length)) {
```
Please put variable on the left side of comparison

```suggestion
  if ((!amf_app_desc_p) || (!nas_msg) || (nas_msg_length == 0)) {
```

ganeshg87 replied:
addressed the comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,212 @@ int send_uplink_nas_registration_complete(
+  rc     = amf_app_handle_uplink_nas_message(
+      amf_app_desc_p, pdu_session_est_req, ue_id, originating_tai);
+
+  return (rc);
```
```suggestion
  return rc;
```

ganeshg87 replied:
addressed the comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,212 @@ int send_uplink_nas_registration_complete(
+  response.result           = 0;
+}
+
+int send_ip_address_response_grpc() {
```
```suggestion
int send_ip_address_response_itti() {
```

ganeshg87 replied:
addressed the comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,212 @@ int send_uplink_nas_registration_complete(
+      AF_INET, "192.168.128.200", response.pdu_address.redirect_server_address);
+}
+
+int send_pdu_session_response_grpc() {
```
```suggestion
int send_pdu_session_response_itti() {
```

ganeshg87 replied:
addressed the comment

ganeshg87 committed:
Addressed review comments(R0)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
Hi @ssanadhya,
I have addressed the review comment and validated registration, pdu setup, dereg with UERANSIM. 
please find the latest UERANSIM logs
[mme_reg_pdu_rel_11_11_21.log](https://github.com/magma/magma/files/7521535/mme_reg_pdu_rel_11_11_21.log)
![image](https://user-images.githubusercontent.com/83060027/141326509-787bfe65-33b5-4dc9-bee6-2c1341b8a96e.png)

executed oai unit test.
![image](https://user-images.githubusercontent.com/83060027/141326452-a6063a99-28d0-44f2-a729-22db4d3ca41f.png)

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -485,11 +486,11 @@ int amf_smf_send(
         dnn_string  = default_dnn;
       } else {
-        dnn_string = msg->dnn.dnn;
+        dnn_string.copy((char*) msg->dnn.dnn, msg->dnn.len - 1);
```
Please address cpplint warning

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h:
```diff
@@ -26,6 +26,13 @@ extern "C" {
 #endif
 #include <memory>
 #include <string>
+#include <lte/protos/session_manager.grpc.pb.h>
```
Use quotes for project headers
```suggestion
#include "lte/protos/session_manager.grpc.pb.h"
```

ganeshg87 replied:
review comment addressed

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h:
```diff
@@ -60,12 +84,11 @@ class AMFClientServicer : public AMFClientServicerBase {
     itti_free_msg_content(message_p);
     free(message_p);
-    return RETURNok;
+    return rc;
```
`rc` is only used once, which can be skipped

```suggestion
    return RETURNok;
```

ganeshg87 replied:
review comment addressed

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,212 @@ int send_uplink_nas_registration_complete(
+  response.pdu_session_id   = 1;
+  response.pdu_session_type = IPv4;
+  response.pti              = 0x01;
+  response.result           = 0;
```
Can we use macros instead of hardcoded numeric values?

ganeshg87 replied:
the above value will change for each unit test case, these values are to compare Hex buffer for UE msg, which will change each registration/PDU.

I agree with your point if any hard-coded values in the AMF application code, I will take this comment as a note.

ssanadhya reviewed (changes requested):
Please address cpp lint warnings too

ganeshg87 committed:
Addressed reiew comments(R1)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
Hi @ssanadhya,
  addressed review comments, cpplint warnings resolved.
  validated the changes with UERANSIM and oai test.

 
![image](https://user-images.githubusercontent.com/83060027/141772061-27630212-be5f-4dbb-bbfb-1dcbd5930568.png)
![image](https://user-images.githubusercontent.com/83060027/141772213-d4d86e77-7cdb-420e-92fa-9980cd3eaa5d.png)
[mme_reg_pdu_15_nov_21.log](https://github.com/magma/magma/files/7538191/mme_reg_pdu_15_nov_21.log)

ganeshg87 committed:
addressed review comment(R2)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
@ulaskozat , @ssanadhya , 
 addressed review comments.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,217 @@ int send_uplink_nas_registration_complete(
+  }
+
+  pdu_session_est_req = blk2bstr(nas_msg, nas_msg_length);
+  memcpy(pdu_session_est_req->data, nas_msg, nas_msg_length);
```
`blk2bstr` is doing the memcpy implicitly. Why do we need an explicit memcpy here?

ganeshg87 replied:
agree. memcpy is not required.
addressed review comment. copied existing test code.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.h:
```diff
@@ -60,6 +60,36 @@ int send_uplink_nas_registration_complete(
+
+int send_pdu_notification_response();
+
+/* Create pdu session establishment  release from ue */
```
```suggestion
/* Create pdu session release from ue */
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,217 @@ int send_uplink_nas_registration_complete(
+  bstring pdu_session_est_req;
+  tai_t originating_tai = {};
+
+  if ((!amf_app_desc_p) || (!nas_msg) || (0 == nas_msg_length)) {
```
Please make the same fix as line 174
```suggestion
  if ((!amf_app_desc_p) || (!nas_msg) || (nas_msg_length == 0)) {
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,217 @@ int send_uplink_nas_registration_complete(
+  if (!ue_context) {
+    return RETURNerror;
+  }
+
```
Why is this part of send response function? It should be part of the test validation.

ganeshg87 replied:
removed above code, it was validating after initial UE Message handling.
```
  /* Check if UE Context is created with correct imsi */
  bool res = false;
  res      = get_ue_id_from_imsi(amf_app_desc_p, imsi64, &ue_id);
  EXPECT_TRUE(res == true);
```

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +164,217 @@ int send_uplink_nas_registration_complete(
+  }
+
+  pdu_session_est_req = blk2bstr(nas_msg, nas_msg_length);
+  memcpy(pdu_session_est_req->data, nas_msg, nas_msg_length);
```
No need for memcpy after blk2bstr.

ganeshg87 replied:
agree.removed memcpy

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -763,8 +763,9 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
 
-  std::string dnn("internet");
-  EXPECT_EQ(dnn, pdu_sess_est_req.dnn.dnn);
+  // std::string dnn("internet");
```
Remove commented code

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -774,7 +775,11 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
   EXPECT_EQ(decode_res, true);
   // build uplinknastransport
-  EXPECT_EQ(dnn, decode_pdu_sess_est_req.dnn.dnn);
+  // EXPECT_EQ(dnn, decode_pdu_sess_est_req.dnn.dnn);
```
Remove commented code and irrelevant comment in line 777

ganeshg87 replied:
addressed review comment

ganeshg87 committed:
addressed review comments(R3)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
@ssanadhya, 
addressed the latest review comments and validated the changes.
1. oai test snap
![image](https://user-images.githubusercontent.com/83060027/142195622-13b5a394-31b5-4125-90ab-63fc258cd994.png)
2. basic reg & pdu & dereg with UERANSIM
![image](https://user-images.githubusercontent.com/83060027/142195678-d9aa7bf4-6e2a-48f1-877f-86d9a079b7d0.png)
[mme_reg_17_11_2021.log](https://github.com/magma/magma/files/7554396/mme_reg_17_11_2021.log)

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +160,217 @@ int send_uplink_nas_registration_complete(
+      amf_ue_context_exists_amf_ue_ngap_id(ue_id);
+  if (!ue_context_p) {
+    return RETURNerror;
+  }
```
This can also be removed from here

ganeshg87 replied:
accessing ue_context_p pointer variable to fetch apn config profile.
> apn_config_profile_t& profile = ue_context_p->amf_context.apn_config_profile;

 if fetching ue context fails, the system should not crash, error reporting is ideal behaviour.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +160,217 @@ int send_uplink_nas_registration_complete(
+  return rc;
+}
+
+/* Create pdu session release  request from ue */
```
```suggestion
/* Create pdu session release request from ue */
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +160,217 @@ int send_uplink_nas_registration_complete(
   return (rc);
 }
 
+/* Create pdu session establishment  request from ue */
```
```suggestion
/* Create pdu session establishment request from ue */
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp:
```diff
@@ -164,6 +160,217 @@ int send_uplink_nas_registration_complete(
+    return RETURNerror;
+  }
+
+  pdu_session_est_req = blk2bstr(nas_msg, nas_msg_length);
```
This is not for establishment, so use a generic variable name
```suggestion
  pdu_session_req = blk2bstr(nas_msg, nas_msg_length);
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:
```diff
@@ -243,4 +256,87 @@ TEST_F(AMFAppProcedureTest, TestDeRegistration) {
+      sizeof(pdu_sess_release_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for pdu session release request from UE */
```
```suggestion
  /* Send uplink nas message for pdu session release complete from UE */
```

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:
```diff
@@ -243,4 +256,87 @@ TEST_F(AMFAppProcedureTest, TestDeRegistration) {
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for pdu session release request from UE */
+  rc = send_uplink_nas_pdu_session_release_request(
```
Since the same function is used to send both pdu session release request/complete, please name the function as `send_uplink_nas_pdu_session_release_message`

ganeshg87 replied:
addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:
```diff
@@ -243,4 +256,87 @@ TEST_F(AMFAppProcedureTest, TestDeRegistration) {
+  rc = send_pdu_notification_response();
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for registration complete response from UE */
```
```suggestion
  /* Send uplink nas message for deregistration complete response from UE */
```

ganeshg87 replied:
addressed review comment

ssanadhya reviewed (changes requested):
The initial reviews have been addressed. Adding some new ones here. In general, it will be a good idea to take a review comment and apply it to the entire PR, wherever applicable. Specifically:
1. Make sure comments convey the logic clearly
2. Function names make clear the intent of the function
3. As @ulaskozat commented on Slack, "we want (macro(...)/var/func(...) == const) and not (const == macro(...)/var/func(...) in condition statements."

ganeshg87 committed:
addressed review comments(R4)

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>
